### DISCUSSION TITLE
Unable to pass CYPRESS_RECORD_KEY as env variable from cypress.env.json

### DISCUSSION THREAD

**Discussion by itkhanz** [cypress-io/cypress#27735] [Questions and Help]
Hi,
I am running locally cypress version 13.0 and trying to pass the `CYPRESS_RECORD_KEY` as environment variable stored in external `cypress.env.json` file  in  the root directoy of the project.

In my cypress.env.json, i have defined the projectId and cypress record key as follows:
```
{
	"projectId": "<ID>",
	"CYPRESS_RECORD_KEY": "<KEY>"
}

```
when I run tests from command line via `npx cypress run --record`, I get the error

```
You passed the --record flag but did not provide us your Record Key.

You can pass us your Record Key like this:

  $ cypress run --record --key <record_key>

You can also set the key as an environment variable with the name: CYPRESS_RECORD_KEY

https://on.cypress.io/how-do-i-record-runs
```

I read the documentation and it states that the CYPRESS_RECORD_KEY can also be passed as environment variable, so I used cypress.env.json to pass it as environment variable. I even checked from cypress test runner app and I can see this environment variable there but running the tests give me above error.
https://docs.cypress.io/guides/cloud/getting-started
https://docs.cypress.io/guides/guides/environment-variables#Option-2-cypressenvjson
https://docs.cypress.io/guides/references/error-messages#You-passed-the---record-flag-but-did-not-provide-us-your-Record-Key

There is also a stack overflow issue created for similar issue, but there is no working solution:
https://stackoverflow.com/questions/74129938/add-cypress-dashboard-record-key-in-cypress-config-file

Could you please look into this issue, and help me pass the record key locally as environment variable from cypress.env.json file?

**Comment by Ingness** [comment]
I have the same problem today. Anyone know the solution?

**Reply by aamrlyman** [reply]
Me too!

**Comment by samelawrence** [comment]
@itkhanz Does test recording succeed or fail if you pass it in as a parameter in your run script?

**Comment by aamrlyman** [comment]
It succeeds if I pass it in as a parameter in my run script. I just never
got it work using the explanation:

You can also set the key as an environment variable with the name:
CYPRESS_RECORD_KEY


On Thu, Feb 22, 2024 at 3:06 PM Sam E. Lawrence ***@***.***>
wrote:

> @itkhanz <https://github.com/itkhanz> Does test recording succeed or fail
> if you pass it in as a parameter in your run script?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cypress-io/cypress/discussions/27735#discussioncomment-8562542>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYQNQMHLJEEQENBIWSSOAE3YU7FQ5AVCNFSM6AAAAAA4JON5AKVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DKNRSGU2DE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Richard Lyman

**Reply by samelawrence** [reply]
I pass the value into my run scripts. I would just do what works and open an issue if you think you have found a legit bug with the other method.

**Comment by jlippold** [comment]
I want to call `npm run test` which records to cloud. 
So I added `"test": "npx cypress run --record"` to package.json
But it complains about the missing CYPRESS_RECORD_KEY

Currently, I am reading /.env like this:

```
const { defineConfig } = require("cypress");
const dotenvPlugin = require("cypress-dotenv");

module.exports = defineConfig({
  e2e: {
    baseUrl: 'http://localhost:8080',
    specPattern: './tests/e2e/**/*.js',
    supportFile: './tests/support/e2e.js',
    screenshotsFolder: './tests/screenshots',
    setupNodeEvents(on, cypressConfig) {
      // inject vars from .env file
      const onlyImportCypressEnv = true;
      const path = 'server/.env';
      const updatedConfig = dotenvPlugin(cypressConfig, { path }, onlyImportCypressEnv);
      return updatedConfig;
    },
  },
});

```

This works for other CYPRESS_ vars, but not the CYPRESS_RECORD_KEY... why not?

If I set the var system wide, it still doesnt take when calling `npx cypress run --record`

echo $CYPRESS_RECORD_KEY shows the var, but cypress doesnt see it?

Where are we supposed to safely store the `CYPRESS_RECORD_KEY` ?

Hardcoded in the package.json? Hardcoded in the cypress.config.js?

Can yall support .env files?

**Comment by rickReachReporting** [comment]
I am sure there are other ways to accomplish this, but I read the .env file in a separate script and executed the script in my package.json.

.env
`CYPRESS_RECORD_KEY = "cypressRecordKeyHere"`
<hr>
 script at project root: execute_script_with_Cypress_Record_key.ts

```
require('dotenv').config();
const cypress = require('cypress');

const executeScriptWithCypressRecordKey = async () => {
  const passedArgs = process.argv[2];
  if (!passedArgs) {
    throw Error('no spec path given');
  }
  cypress.run({
    spec: passedArgs,
    record: true,
    key: process.env.CYPRESS_RECORD_KEY,
  });
};

executeScriptWithCypressRecordKey();
```
<hr>
package.json

```
"scripts": {
    "functional": "node execute_script_with_Cypress_Record_key.ts ./cypress/e2e/Reach_Functional",
}
```

**Comment by jennifer-shehane** [comment]
This is working as intended. `CYPRESS_RECORD_KEY` is an OS-level environment variable, not a Cypress (test) env var, so it can’t be set via `cypress.env.json` or the `env` config. Use one of these instead:

```
# inline flag
npx cypress run --record --key <record_key>

# or an OS env var (locally, or as a CI secret)
export CYPRESS_RECORD_KEY=<record_key>
npx cypress run --record
```

I’ve opened a PR to make this clearer in our docs.
